drivers/espi/espi_emul.c

```diff
@@ -21,170 +21,477 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
-	    ((evt->evt_type & ESPI_BUS_PERIPHERAL_NOTIFICATION) &&
-	     !(data->cfg.channel_caps & ESPI_CHANNEL_PERIPHERAL))) {
-		return -EIO;
+	if (data->kbc8042_irq_enabled == false || data->kbc8042_obe_irq_pending == false) {
+		return;
 	}
 
-	espi_send_callbacks(&data->callbacks, dev, *evt);
+	data->kbc8042_obe_irq_pending = false;
+
+	ev.evt_type = ESPI_BUS_PERIPHERAL_NOTIFICATION;
+	ev.evt_details = ESPI_PERIPHERAL_8042_KBC;
+
+	kbc_evt = (struct espi_evt_data_kbc *)&ev.evt_data;
+	kbc_evt->evt = HOST_KBC_EVT_OBE;
+	kbc_evt->data = 0;
+	kbc_evt->type = 0;
+
+	espi_send_callbacks(&data->callbacks, dev, ev);
+}
+#endif
+
+/** Controller API **/
+
+static int espi_emul_api_config(const struct device *dev, struct espi_cfg *cfg)
+{
+	struct espi_emul_data *data = dev->data;
+
+	data->channels_enabled = cfg->channel_caps;
 
 	return 0;
 }
 
-static bool espi_emul_get_channel_status(const struct device *dev, enum espi_channel ch)
+static bool espi_emul_api_get_channel_status(const struct device *dev, enum espi_channel ch)
 {
 	struct espi_emul_data *data = dev->data;
 
-	return (data->cfg.channel_caps & ch);
+	return (data->channels_enabled & ch);
 }
 
-static int espi_emul_read_lpc_request(const struct device *dev, enum lpc_peripheral_opcode op,
-				      uint32_t *data)
+/* Logical Channel 0 APIs */
+static int espi_emul_api_read_request(const struct device *dev, struct espi_request_packet *req)
 {
-	const struct emul_espi_device_api *api;
-	struct espi_emul *emul;
-	struct espi_emul_data *emul_data = dev->data;
+	return -EIO;
+}
 
-	ARG_UNUSED(data);
+static int espi_emul_api_write_request(const struct device *dev, struct espi_request_packet *req)
+{
+	return -EIO;
+}
 
-	if (!(emul_data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
-		LOG_ERR("bad channel vwire");
-		return -EINVAL;
-	}
+static int espi_emul_api_lpc_read_request(const struct device *dev, enum lpc_peripheral_opcode op,
+					  uint32_t *value)
+{
+	struct espi_emul_data *data = dev->data;
+	struct espi_emul *emul = data->emul;
 
-	emul = espi_emul_find(dev, EMUL_ESPI_HOST_CHIPSEL);
-	if (!emul) {
-		LOG_ERR("espi_emul not found");
-		return -ENOTSUP;
+	if ((data->channels_enabled & ESPI_CHANNEL_PERIPHERAL) == 0) {
+		return -EIO;
 	}
 
-	__ASSERT_NO_MSG(emul->api);
-	api = emul->api;
-
 	switch (op) {
-#ifdef CONFIG_ESPI_PERIPHERAL_ACPI_SHM_REGION
-	case EACPI_GET_SHARED_MEMORY:
-		__ASSERT_NO_MSG(api->get_acpi_shm);
-		*data = (uint32_t)api->get_acpi_shm(emul->target);
+#ifdef CONFIG_ESPI_EMUL_KBC8042
+	case E8042_OBF_HAS_CHAR:
+		*value = (data->kbc8042_status & KBC8042_STATUS_OBF) != 0 ? 1 : 0;
+		break;
+	case E8042_IBF_HAS_CHAR:
+		*value = (data->kbc8042_status & KBC8042_STATUS_IBF) != 0 ? 1 : 0;
+		break;
+	case E8042_READ_KB_STS:
+		*value = data->kbc8042_status;
 		break;
 #endif
+	/* ACPI status transactions */
+	case EACPI_READ_STS:
+		*value = data->acpi_status;
+		break;
 	default:
-		return -EINVAL;
+		__ASSERT_NO_MSG(emul != NULL);
```

Comment:
The other functions registered with `struct espi_driver_api` return -EIO if the host emulator isn't provided.
Is it valid to use this peripheral emulator without a host emulator?   If the host emulator is required, you could move this check into `espi_emul_init()` or `espi_emul_api_config()`

Reply:
Technically it's correct, you can have an eSPI peripheral which isn't connected to any host.